benchadapt/python/benchadapt/result.py

```diff
@@ -63,18 +63,45 @@ class BenchmarkResult:
         For benchmarks run on a cluster, information about the cluster
     context : Dict[str, Any]
         Should include ``benchmark_language`` and other relevant metadata like compiler flags
-    github : Dict[str, Any]
-        Keys: ``repository`` (in the format ``org/repo``), ``commit``, and ``pr_number``.
-        If this is a benchmark on the default branch, you may leave out ``pr_number``.
-        If it's a non-default-branch & non-PR commit, you may supply the branch name to
-        the optional ``branch`` key in the format ``org:branch``.
+    github : Union[Optional[Dict[str, Any]], Literal["inspect_git_in_cwd"]
 
-        By default, metadata will be obtained from ``CONBENCH_PROJECT_REPOSITORY``,
-        ``CONBENCH_PROJECT_COMMIT``, and ``CONBENCH_PROJECT_PR_NUMBER`` environment variables.
-        If any are unset, a warning will be raised.
+        A dictionary containing GitHub-flavored commit information.
 
-        Advanced: if you have a locally cloned repo, you may explicitly supply ``None``
-        to this argument and its information will be scraped from the cloned repo.
+        Allowed values: `None`, no value, a dictionary.
+
+        Not passing an argument upon dataclass construction results in inspection
+        of the environment variables ``CONBENCH_PROJECT_REPOSITORY``,
+        ``CONBENCH_PROJECT_COMMIT``, and ``CONBENCH_PROJECT_PR_NUMBER``.
+        If any of those are not set, a warning is raised.
+
+        Passing `None` explicitly defines that this result is not associated
+        with any commit.
+
+        Not associating a benchmark result with commit information has special,
+        limited purpose (pre-merge benchmarks, testing). It generally means
+        that this benchmark result will not be considered for time series
+        analysis along a commit tree.
+
+        If passed a dictionary, it should have the following keys:
+        - required: ``repository`` (string, in the format ``https://github.com/<org>/<repo>``)
+        - required: ``commit`` (string, the full commit hash)
+        - recommended: ``pr_number`` (stringified integer) or ``branch`` (string)
+
+        If the benchmark was run gainst the default branch, you may leave
+        out ``pr_number`` and ``branch``.
+
+        If it was run on a GitHub pull request branch, you should provide the
+        ``pr_number``.
+
+        If it was run on a non-default branch and a non-PR commit, you may
+        supply the branch name via the ``branch`` set to a value of the format
+        ``org:branch``.
```

Comment:
This should also mention that `"pr_number": None` is the way to specify a commit is on the default branch

Reply:
Yes.
For the record, documented here: https://github.com/conbench/conbench/blob/6c6f3a441a07a0e3a4ce3f6f5ca46e0ff3ca8d3e/conbench/entities/run.py#L568

Btw there is the ambiguity of this being defined as marshmallow.fields.Integer() and here I think we send it as string, may be fine, maybe marshmallow is OK with that.

In this module we should use the same type as here: https://github.com/conbench/conbench/blob/6c6f3a441a07a0e3a4ce3f6f5ca46e0ff3ca8d3e/conbench/entities/commit.py#L43

(not saying the type is perfect, but code-wise we should use the _same_ type definition (monorepo allows us to!).

Reply:
Well.

> is the way

The API spec says that "null" is the way, but probably allows both (not-set, set to null).

And that's interesting because in JSON there is a semantic difference between not set and set-to-null.

This depends on marshmallow schema validation.

Need to sharpen the API spec in that regard.

Reply:
I now also remember that the famous test you added recently with `@pytest.mark.parametrize("pr_number", [12345678, "12345678", None, "", "<absent>"])` covers all this. But yeah, this needs to be textually specified, for our own sake.

Reply:
The more I think about this right now, and the more I go back and forth between the different places in code where we talk about this and deal with, the clearer it becomes:

There needs to be a _central place_ for textual specification that all other places refer to. I keep typing the same sentences in different places, and over time they diverge.

Reply:
I took your feedback serious and added a commit, working a little more on the HTTP/JSON API spec. Also worked on this text here again.

Future: add typedict / dataclass /pydantic datatype for this dict, and perform runtime schema validation.